File: sova/api/forms/react_py/react_py.py
# -*- coding: utf-8 -*-
'''
Created on 13 февр. 2020 г.

@author: aon24
'''
# *** *** ***

from .. fm_svg.svgTools import svgToPy
from .. formTools import style, _div, _field, _h2, _h3, _btnD, _a

from math import *
import json

jsUrl = 'jsv?api/forms/react_py/react_py.js'
cssUrl = 'jsv?api/forms/react_py/react_py.css'

title = 'react-py'

# *** *** ***

notes = [
'''
<tspan x="10" dy="0.5em" stroke-width="0" font-weight="normal" font-size="14px">
В сложных системах нет мелочей.</tspan>
<tspan x="8" dy="1.5em" stroke-width="0" font-weight="normal" font-size="14px">
При постоянной работе с програм-</tspan>
<tspan x="12" dy="1.5em" stroke-width="0" font-weight="normal" font-size="14px">
мой задержки даже в 1-2 секунды</tspan>
<tspan x="30" dy="1.5em" stroke-width="0" font-weight="normal" font-size="14px">
раздражают. Пример работы</tspan>
<tspan x="40" dy="1.5em" stroke-width="0" font-weight="normal" font-size="14px">
с кэшируемым списком </tspan>
<tspan x="60" dy="1.5em" stroke-width="0" font-weight="normal" font-size="14px">
из 1000 объектов.</tspan>
''',

'''Фреймворк React-py (Sova) это:
<tspan x="30" dy="1.5em">- мост между Python и React</tspan>
<tspan x="30" dy="1.5em">- открытый код</tspan>
<tspan x="30" dy="1.5em">- меньше 300 Кб JS</tspan>
<tspan x="45" dy="1.5em" stroke-width="0" font-weight="normal" font-size="14px">
Python 3.7 / React 16.12</tspan>
<tspan x="65" dy="1.5em" stroke-width="0" font-weight="normal" font-size="14px">
Linux, Windows 7/10 </tspan>
''',

'''
<tspan x="10" dy="0.5em" stroke-width="0" font-weight="normal" font-size="14px">
В сложных системах нет мелочей.</tspan>
<tspan x="8" dy="1.5em" stroke-width="0" font-weight="normal" font-size="14px">
Грамотно организованная система</tspan>
<tspan x="12" dy="1.5em" stroke-width="0" font-weight="normal" font-size="14px">
логгирования облегчает отладку</tspan>
<tspan x="22" dy="1.5em" stroke-width="0" font-weight="normal" font-size="14px">
и сопровождение приложения.</tspan>
<tspan x="40" dy="1.5em" stroke-width="0" font-weight="normal" font-size="14px">
Недаром модуль назван</tspan>
<tspan x="100" dy="1.5em">first.py</tspan>
''',

'''Разработка форм на 2-х экранах
<tspan fill="#048" x="5" dy="1.5em" stroke-width="0" font-weight="normal" font-size="14px">
При редактировании формы</tspan>
<tspan fill="#048" x="10" dy="1.5em" stroke-width="0" font-weight="normal" font-size="14px">
изменения будут отображаться</tspan>
<tspan fill="#048" x="20" dy="1.5em" stroke-width="0" font-weight="normal" font-size="14px">
в реальном времени на всех</tspan>
<tspan fill="#048" x="30" dy="1.5em" stroke-width="0" font-weight="normal" font-size="14px">
подключенных устройствах.</tspan>
<tspan fill="#048" x="50" dy="1.5em" stroke-width="0" font-weight="normal" font-size="14px">
Синхронизация через</tspan>
<tspan fill="#048" x="90" dy="1.5em" stroke-width="1" font-weight="normal" font-size="14px">
WebSocket.</tspan>
''',

'''Пример многооконного прило-
<tspan x="15" dy="2.5em">жения. Вы можете двигать</tspan>
<tspan x="35" dy="2.5em">окна и открывать окна</tspan>
<tspan x="60" dy="2.5em">в новых вкладках.</tspan>
''',

'''
Пример WEB-приложения для
<tspan x="12" dy="1.5em">
смартфона. Заказчик оценил</tspan>
<tspan x="22" dy="1.5em">
на 5+.</tspan>
<tspan fill="#048" x="80" stroke-width="0" font-weight="normal" font-size="14px">
Установка на телефон:</tspan>
<tspan fill="#048" x="30" dy="1.5em" stroke-width="0" font-weight="normal" font-size="14px">
открыть страницу в браузере </tspan>
<tspan fill="#048" x="40" dy="1.5em" stroke-width="0" font-weight="normal" font-size="14px">
и добавить ее на главный</tspan>
<tspan fill="#048" x="50" dy="1.5em" stroke-width="0" font-weight="normal" font-size="14px">
экран.</tspan>
''',
]

def priz():
    cx, cy = 400, 400
    r = 350
    p1 = p2 = p3 = ''
    pi3, pi6 = pi/3, pi/6
    color = ['#efd', '#eee', '#fde', '#edf', '#dfe', '#def']
    for  i in range(6):
        ccx = cx+(r-160)*cos(pi6+pi3*i)
        ccy = cy+(r-160)*sin(pi6+pi3*i)
        tx = cx+(r-10)*cos(pi*0.065+pi3*i)
        ty = cy+(r-10)*sin(pi*0.065+pi3*i)
        tx2 = cx+(r-10)*cos(pi*0.275+pi3*i)
        ty2 = cy+(r-10)*sin(pi*0.275+pi3*i)
        refs = ['/newdoc?&amp;themes', '/newdoc?&amp;home', '/newdoc?&amp;lm',
                '/home-ru', '/newdoc?&amp;mp_spa', '/newdoc?_&amp;schedule']
        text = [
                '___Кэшируемые списки______',
                '_____ Документация _______',
                '_________ Logging ________',
                '_____ Layout manager______',
                '____ Multi-window SPA_____',
                '_____ Почасовая ареда_____',
                ]

        #'/docopen?SOVA/SCHEDULE&amp;20202020202020202020202020202020&amp;edit'
        p1 += f'''
<a href="{refs[i]}" target="_blank">
    <circle cx="{ccx}" cy="{ccy}" r="210" fill="{color[i]}"/>
    <defs><path id="txt1{i}"  d="M {tx},{ty} A 160,160 0 0 1 {tx2},{ty2}"/></defs>
    <defs><path id="txt2{i}"  d="M {tx},{ty} L {tx2},{ty2}"/></defs>
    <text x="0" y="0" style="stroke: #00f; font: normal 20px Times">
      <textPath xlink:href="#txt1{i}">{text[i]}</textPath>
    </text>
    <text x="{0}" y="{0}" style="stroke: #00f; font: normal 20px Times">
        <textPath xlink:href="#txt2{i}"> ________Открыть______</textPath>
    </text>
</a>
'''
    for i in range(6):
        stroke = ['#555', '#555', '#555', '#048', '#00f', '#048']
        tx2 = cx+(r-20)*cos(pi*0.28+pi3*i) if i+3 < 3 else cx+(r-35)*cos(pi*0.02+pi3*i) 
        ty2 = cy+(r-20)*sin(pi*0.28+pi3*i) if i+3 < 3 else cy+(r-35)*sin(pi*0.02+pi3*i)
        tx3 = cx+(r-20)*cos(pi*0.13+pi3*i) if i+3 < 3 else cx+(r-35)*cos(pi*0.312+pi3*i)
        ty3 = cy+(r-20)*sin(pi*0.13+pi3*i) if i+3 < 3 else cy+(r-35)*sin(pi*0.312+pi3*i)
        s = f'fill="{color[i]}" stroke="#fff" stroke-width="1"'
        p1 += f'''<polygon {s} points="{cx+r*cos(pi3*i)},{cy+r*sin(pi3*i)} {cx+r*cos(pi3*(i+1))},{cy+r*sin(pi3*(i+1))} {cx},{cy}"/>
<defs><path id="txt3{i}"  d="M {tx2},{ty2} L {tx3},{ty3}"/></defs>
<text x="{0}" y="{0}" stroke="{stroke[i]}" style="font: normal 16px Arial" letter-spacing="1.2">
  <textPath xlink:href="#txt3{i}">{notes[i]}</textPath>
</text>
'''
    
    p2 += f'''
<circle cx="{cx}" cy="{cy}" r="150" fill="#fff" stroke="#048" stroke-width="1"/>
<circle cx="{cx}" cy="{cy}" r="110" fill="#fed" stroke="#048" stroke-width="1"/>
'''
    
    p2 += f'''
<defs><path id="script1" d="M {cx-126},{cy}  A 126,126 0 0 1 {cx+126},{cy}"/></defs>
<defs><path id="script2" d="M {cx-131},{cy}  A 131,131 0 0 0 {cx+131},{cy}"/></defs>
<defs><path id="python" d="M {cx-80},{cy}  A 80,80 0 0 1 {cx+80},{cy}"/></defs>
<defs><path id="react"  d="M {cx-100},{cy}  A 100,100 0 0 0 {cx+100},{cy}"/></defs>
<text x="20" y="0" style="stroke: #048; font: normal 15px Times" letter-spacing="1.5">
  <textPath xlink:href="#script1">Все приложения разработаны на Python 3.7</textPath>
</text>
<text x="30" y="0" style="stroke: #048; font: normal 15px Times" letter-spacing="1.5">
  <textPath xlink:href="#script2">Javascript общий для всех приложений</textPath>
</text>
<text x="30" y="0" style="stroke: #048; font: normal 30px Times">
  <textPath xlink:href="#python">R e a c t - p y</textPath>
</text>
<text x="50" y="0" style="stroke: #048; font: normal 20px Times">
  <textPath xlink:href="#react">P y t h o n  -  R e a c t </textPath>
</text>'''

    for i in range(12):
        p3 += '<polygon points="0,0 0,0 0,0 0,0" stroke-width="1"/>'

    svg = f'''
<svg width="{2*cx+20}" height="{2*cy}">
    <svg width="{2*cx+20}" height="{2*cy}"><g>{p1}</g></svg>
    <svg width="{2*cx+20}" height="{2*cy}">{p2}</svg>
    <svg width="{2*cx+20}" height="{2*cy}">{p3}</svg>
</svg>'''

    return [ svgToPy(svg) ]

# *** *** ***

def page(dbAlias, mode, userName, multiPage):
    def descr(name, width):
        ds = [
            ['Пример многооконного приложения',
             '''Фремворк основан на принципах объектно-оринтированного программирования. Объекты генерируют и обрабатывают события в стиле <<B+ООП>>, что гораздо проще и понятней, чем <<B:Redux>>. 
             В примере вы можете двигать окна и открывать окна в новых вкладках.\n''',
             '/newdoc?&mp_spa'],
            
            ['Пример WEB-приложения для смартфона',
             '''Заказчик оценил на <<R+5+>>. Установка на телефон:
                - открыть страницу в браузере
                - добавить ее на главный экран.''',
             '/newdoc?_&schedule', '/home-ru'],
                 
            ['Разработка форм на 2-х экранах',
             '''При редактировании формы изменения будут отображаться в реальном времени на всех подключенных устройствах.
             Синхронизация через <<B+WebSocket>>.''',
            '/home-ru'],
        
            ['',
             '''В сложных системах нет мелочей. При постоянной работе с программой задержки даже в 1-2 секунды раздражают.
             Пример работы с кэшируемым списком из 1000 объектов.
             ''',
            '/newdoc?&themes'],

            ['Журнал работы приложения', '''
            В сложных системах нет мелочей. Грамотно организованная система логгирования облегчает отладку и сопровождение приложения.
            Недаром модуль назван <<B+first.py>>''',           
            '/newdoc?&lm'],

            ['','''Описание несколько устарело. Фреймворк постоянно развивается и меняется в сторону упрощения.
            <<C+Пример>>. <<B:React-py>> для прорисовки должен получить словарь:
            \xa0
{'<<R:_teg>>': '<<G+div>>', '<<R:text>>': '<<G+Hello>>-', '<<R:attributes>>': {'<<R:className>>': '<<G+h2>>'},
    \xa0\xa0\xa0\xa0'<<R:children>>': [{'<<R:_teg>>': '<<G+a>>', '<<R:text>>': '<<G+Привет>>',
    \xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0'<<R:attributes>>': {'<<R:href>>': <<G+'http://ya.ru>>', '<<R:style>>': {'<<R:color>>': '<<G+red>>'}}}]
}            
            тот же словарь можно создать с помощью элементарных функций:
\xa0
<<B+_h2>>('<<G+Hello>>-', <<R:children>>=[<<B+_a>>('<<G+Привет>>', <<R:href>>=<<G+'http://ya.ru>>', **<<B+style>>(<<R:color>>='<<G+red>>'))]),
''',
            '/newdoc?&home'],
        ]
        
        return [ _div(name=f'{name}{i+1}', className='descr', **style(width=width),
                    children=[
                        _h2(ds[i][0]),
                        _div(ds[i][1], br='p'),
                        _h2('Hello-', children=[_a('Привет', href='http://ya.ru', **style(color='red'))]) if i == 5 else None,
                        _btnD('Открыть', 'xopen', ds[i][2], className='svTop', **style(width=100, margin='10px auto'))
                    ]
                ) for i in range(6)]

        
    tit = [
        '1. Multi-window SPA',
        '2. Почасовая ареда',
        '3. Layout manager',
        '4. Кэшируемые списки',
        '5. Logging',
        '6. Документация',
    ]

    return _div(
        **style(overflow='auto'),
        children = [
            _h2('Фрейморк React-py ', **style(color='blue')),
            _h3('''7 примеров, разработаных на Python 3.7.
                Во всех приложениях использован один
                Javascript (всего 260 Кб)''', **style(color='#048'), br=1),
            
            _field('chMode_fd', 'chb3', dropList=['презентация|p', 'описание|n'], **style(width=280, font='normal 24px Arial', margin='10px auto'), noEmpty=1),

            _div(name='plain1', className='tbl', children=
                [_btnD(tit[i], f'app1{i+1}', className='nosel', **style(borderBottomWidth=0)) for i in range(3, 6)],
                ),
            _div(name='plain1', className='tbl', **style(width=660, background='#eee'), children=
                [_btnD(tit[i], f'app{i+1}', className=f'{"nosel" if i else "sel"}') for i in range(3)],
                ),

            _div(name='plain2', className='tbl', children=
                [_btnD(tit[i], f'app1{i+1}', className='nosel', **style(borderBottomWidth=0)) for i in range(3)],
                ),
            _div(name='plain2', className='tbl', **style(width=660, background='#eee'), children=
                [_btnD(tit[i], f'app{i+1}', className='nosel') for i in range(3, 6)],
                ),

            _div(name='plain_1', className='tbl', **style(width=360,) , children=
                [_btnD(tit[i], f'app_1{i+1}', className='nosel', **style(borderBottomWidth=0)) for i in range(3, 6)],
                ),
            _div(name='plain_1', className='tbl', **style(width=360, background='#eee'), children=
                [_btnD(tit[i], f'app_{i+1}', className=f'{"nosel" if i else "sel"}') for i in range(3)],
                ),
            _div(name='plain_2', className='tbl', **style(width=360,) , children=
                [_btnD(tit[i], f'app_1{i+1}', className='nosel', **style(borderBottomWidth=0)) for i in range(3)],
                ),
            _div(name='plain_2', className='tbl', **style(width=360, background='#eee'), children=
                [_btnD(tit[i], f'app_{i+1}', className='nosel') for i in range(3, 6)],
                ),


            *descr('descr', 660),
            *descr('descr_', 360),

            _field('polygon_fd', 'json', **style(width=400, margin='auto', textAlign='center')),

            _field('screen_fd', 'chb3', name='priz', dropList=['смартфон|s', 'компьютер|p'], **style(width=250, margin='10px auto'), noEmpty=1),
        ])
    
# *** *** ***

def queryOpen(d, mode, ground):
    d._polygon = json.dumps(priz(), ensure_ascii=False)
    d.chMode_fd = 'p'

# *** *** ***

File: sova/api/forms/react_py/react_py.js
// *** *** ***

let pi_3 = [0, -1*Math.PI/3, -2*Math.PI/3, -Math.PI, -4*Math.PI/3, -5*Math.PI/3, 0];
let _grad = Math.PI/180;
let [_d, _y, r, _u, _doc, _oPol, _setI, _cx, _cy, _u2] = [280, 75, 150, 0.0, null, null, null, 400, 400, 0];

let _selectedApp;

// *** *** ***

window.sovaActions = window.sovaActions || {};
window.sovaActions.react_py = {
    init: doc => {
        doc.fieldValues['SCREEN_FD'] = doc.util.anyMobile ? 's' : 'p';
        _selectedApp = doc.util.anyMobile ? 'BTN_APP_1' : 'BTN_APP1';
    },
	init2: doc => {
//		console.log(doc.util.anyMobile);
		_doc = doc;
		_oPol = JSON.parse(doc.getField('_polygon'));
                _oPol[0].attributes = _doc.util.anyMobile ?
                    {width: "820", height: "800", viewBox: "0 0 1640 1600"}
                    :
		    {width: "820", height: "800", viewBox: "0 0 1640 1600"};
		_oPol[0].children[0].children[0].attributes = {};
        if ( doc.getField('CHMODE_FD') === 'p' ) {
            doc.setField('polygon_fd', _oPol);
            _setI = setInterval( __, 25);
        }
	},
	
    //*** *** ***

    recalc: {
		CHMODE_FD: (doc, label) => {
            if (label === 'n') { // normal
                clearInterval(_setI);
                doc.setField('polygon_fd', null);
            }
            else {
                _setI = setInterval( __, 50);
            }
            doc.forceUpdate();
		},
        SCREEN_FD: (doc, newValue) => {
            doc.util.anyMobile = newValue !== 'p'; // p - значит компьютер
            runChApp(doc, doc.util.anyMobile ? '_1' : '1');
            doc.forceUpdate();
        },
    },
    
    // *** *** ***
   
    hide: {
        plain1: doc => doc.getField('CHMODE_FD') !== 'n' || _selectedApp > 'BTN_APP3' || doc.util.anyMobile,
        plain2: doc => doc.getField('CHMODE_FD') !== 'n' || _selectedApp <= 'BTN_APP3' || doc.util.anyMobile,
        
        plain_1: doc => doc.getField('CHMODE_FD') !== 'n' || _selectedApp > 'BTN_APP_3' || !doc.util.anyMobile,
        plain_2: doc => doc.getField('CHMODE_FD') !== 'n' || _selectedApp <= 'BTN_APP_3' || !doc.util.anyMobile,
        
        descr1: doc => doc.getField('CHMODE_FD') !== 'n' || _selectedApp != 'BTN_APP1' || doc.util.anyMobile,
        descr2: doc => doc.getField('CHMODE_FD') !== 'n' || _selectedApp != 'BTN_APP2' || doc.util.anyMobile,
        descr3: doc => doc.getField('CHMODE_FD') !== 'n' || _selectedApp != 'BTN_APP3' || doc.util.anyMobile,
        descr4: doc => doc.getField('CHMODE_FD') !== 'n' || _selectedApp != 'BTN_APP4' || doc.util.anyMobile,
        descr5: doc => doc.getField('CHMODE_FD') !== 'n' || _selectedApp != 'BTN_APP5' || doc.util.anyMobile,
        descr6: doc => doc.getField('CHMODE_FD') !== 'n' || _selectedApp != 'BTN_APP6' || doc.util.anyMobile,

        descr_1: doc => doc.getField('CHMODE_FD') !== 'n' || _selectedApp != 'BTN_APP_1' || !doc.util.anyMobile,
        descr_2: doc => doc.getField('CHMODE_FD') !== 'n' || _selectedApp != 'BTN_APP_2' || !doc.util.anyMobile,
        descr_3: doc => doc.getField('CHMODE_FD') !== 'n' || _selectedApp != 'BTN_APP_3' || !doc.util.anyMobile,
        descr_4: doc => doc.getField('CHMODE_FD') !== 'n' || _selectedApp != 'BTN_APP_4' || !doc.util.anyMobile,
        descr_5: doc => doc.getField('CHMODE_FD') !== 'n' || _selectedApp != 'BTN_APP_5' || !doc.util.anyMobile,
        descr_6: doc => doc.getField('CHMODE_FD') !== 'n' || _selectedApp != 'BTN_APP_6' || !doc.util.anyMobile,
        
        priz: doc => doc.getField('CHMODE_FD') !== 'n',
    },
    cmd: {
        app1: doc => runChApp(doc, '1'), app11: doc => runChApp(doc, '1'),
        app2: doc => runChApp(doc, '2'), app12: doc => runChApp(doc, '2'),
        app3: doc => runChApp(doc, '3'), app13: doc => runChApp(doc, '3'),
        app4: doc => runChApp(doc, '4'), app14: doc => runChApp(doc, '4'),
        app5: doc => runChApp(doc, '5'), app15: doc => runChApp(doc, '5'),
        app6: doc => runChApp(doc, '6'), app16: doc => runChApp(doc, '6'),

        app_1: doc => runChApp(doc, '_1'), app_11: doc => runChApp(doc, '_1'),
        app_2: doc => runChApp(doc, '_2'), app_12: doc => runChApp(doc, '_2'),
        app_3: doc => runChApp(doc, '_3'), app_13: doc => runChApp(doc, '_3'),
        app_4: doc => runChApp(doc, '_4'), app_14: doc => runChApp(doc, '_4'),
        app_5: doc => runChApp(doc, '_5'), app_15: doc => runChApp(doc, '_5'),
        app_6: doc => runChApp(doc, '_6'), app_16: doc => runChApp(doc, '_6'),
    }
};

let runChApp = (doc, s) => {
    doc.setField(_selectedApp, {className: 'nosel'});
    doc.setField('BTN_APP' + s, {className: 'sel'});
    _selectedApp = 'BTN_APP' + s;
    doc.forceUpdate();
}
// *** *** ***

let _priz = ugr => {
	_oPol[0].attributes = _doc.util.anyMobile ?
		{width: "820", height: "800", viewBox: "0 0 1640 1600"}
		:
		{width: "820", height: "800", viewBox: "0 0 1640 1600"};
	
	if (_y > 0) _y -= _y > 1 ? 1 : 0.12;
	 {
		if (r > 45) {
			_d -= 2;
			r--;
		}
	}

    let ls = [];
    let [x1, y1, x11, y11] = [ [], [], [], [] ];
    let ug = ugr*_grad;

	pi_3.forEach( (u, i) => {
        let [x,y] = [Math.cos(u+ug), Math.sin(u+ug)];
        [x1[i], y1[i]] = [0.87*(x-y), 0.25*(-x-y)];
		if (_y > 0) {
			[x,y] = [Math.cos(-u-ug), Math.sin(-u-ug)];
			[x11[i], y11[i]] = [0.87*(-x+y), 0.25*(x+y)];
		}
	});

	if (_y <= 0)
		[x11, y11] = [x1, y1];

    for (let i = 0; i < 6; i++) {
        ls[i+6] = [_cx+r*x1[i], _cy+r*y1[i]-_y, _cx, _cy-_d-_y, _cx+r*x1[i+1], _cy+r*y1[i+1]-_y, `rgba(176,${255*6*0.0},${255*6*0.0}, ${0.7})`, '#faa'];
		rand = Math.random() / 10.;
		ls[i]   = [_cx+r*x11[i], _cy+r*y11[i]+_y, _cx, _cy+_d+_y, _cx+r*x11[i+1], _cy+r*y11[i+1]+_y, `rgba(176,${255*6*0.15},${255*6*0.15}, ${0.7})`, '#fff'];
//		ls[i]   = [_cx+r*x1[i], _cy+r*y1[i]+y, _cx, 680+y, _cx+r*x1[i+1], _cy+r*y1[i+1]+y, 'rgba(0,35,74, ${0.7})'];

	}
	return ls;
};

// *** *** ***

const __ = _ => {
    if ( !_oPol )
        return;
	let ls = _priz(_u);
	let ir = 4 + Math.floor((_u+25.)/60.);
	let ib = 4 + Math.floor((_u-(_y > 0 ? 90. : 0))/60.);
	if ( ir >= 6 )
		ir -= 6;

	if ( ib >= 6 )
		ib -= 6;

	if ( r <= 45 ) {
		_sec = 0;
		_u2 -= _doc.util.anyMobile ? 0.2 : 0.1;
		if (_u2 <= -360)
			u2 = 0;
		_oPol[0].children[0].children[0].attributes.transform = `rotate(${_u2} ${_cx} ${_cy})`
	}
	
	_oPol[0].children[2].children.forEach( (it,j) => {
		let i = j;
		if ( j < 6 ) {
			i += ib;
			if ( i >= 6 )
				i -= 6;
		}
		else {
			i += ir;
			if ( i >= 12 )
				i -= 6;
		}
		it.attributes.points = `${ls[i][0]},${ls[i][1]} ${ls[i][2]},${ls[i][3]} ${ls[i][4]},${ls[i][5]}`;
		it.attributes.fill = ls[i][6];
		it.attributes.stroke = ls[i][7];
	});
	_doc.setField('polygon_fd', _oPol, 0);
	_u += 0.5;
	if (_u >= 360)
		_u = 0.0;
	if ( r <= 45 ) {
		clearInterval(_setI);
		_setI = setInterval( __, 50);
	}
	else
		_u += 1;
}

















